xcm/src/v2/multilocation.rs

```diff
@@ -467,89 +468,109 @@ const MAX_JUNCTIONS: usize = 8;
-	X7(Junction, Junction, Junction, Junction, Junction, Junction, Junction),
+	X7(Arc<[Junction; 7]>),
 	/// A relative path comprising 8 junctions.
-	X8(Junction, Junction, Junction, Junction, Junction, Junction, Junction, Junction),
+	X8(Arc<[Junction; 8]>),
+}
+
+// TODO: Remove this once deriving `MaxEncodedLen` on `Arc`s works.
+impl MaxEncodedLen for Junctions {
+	fn max_encoded_len() -> usize {
+		<[Junction; 8]>::max_encoded_len().saturating_add(1)
+	}
+}
+
+macro_rules! impl_junction {
+	($count:expr, $variant:ident) => {
+		impl From<[Junction; $count]> for Junctions {
+			fn from(junctions: [Junction; $count]) -> Self {
+				Self::$variant(Arc::new(junctions))
+			}
+		}
+		impl PartialEq<[Junction; $count]> for Junctions {
+			fn eq(&self, rhs: &[Junction; $count]) -> bool {
+				self.as_slice() == rhs
+			}
+		}
+	};
+}
+
+impl_junction!(1, X1);
+impl_junction!(2, X2);
+impl_junction!(3, X3);
+impl_junction!(4, X4);
+impl_junction!(5, X5);
+impl_junction!(6, X6);
+impl_junction!(7, X7);
+impl_junction!(8, X8);
+
+pub struct JunctionsIterator {
+	junctions: Junctions,
+	range: Range<usize>,
 }
 
-pub struct JunctionsIterator(Junctions);
 impl Iterator for JunctionsIterator {
 	type Item = Junction;
 	fn next(&mut self) -> Option<Junction> {
-		self.0.take_first()
+		self.junctions.at(self.range.next()?).cloned()
 	}
 }
 
 impl DoubleEndedIterator for JunctionsIterator {
 	fn next_back(&mut self) -> Option<Junction> {
-		self.0.take_last()
+		self.junctions.at(self.range.next_back()?).cloned()
 	}
 }
 
 pub struct JunctionsRefIterator<'a> {
 	junctions: &'a Junctions,
-	next: usize,
-	back: usize,
+	range: Range<usize>,
 }
 
 impl<'a> Iterator for JunctionsRefIterator<'a> {
 	type Item = &'a Junction;
 	fn next(&mut self) -> Option<&'a Junction> {
-		if self.next.saturating_add(self.back) >= self.junctions.len() {
-			return None
-		}
-
-		let result = self.junctions.at(self.next);
-		self.next += 1;
-		result
+		self.junctions.at(self.range.next()?)
 	}
 }
 
 impl<'a> DoubleEndedIterator for JunctionsRefIterator<'a> {
 	fn next_back(&mut self) -> Option<&'a Junction> {
-		let next_back = self.back.saturating_add(1);
-		// checked_sub here, because if the result is less than 0, we end iteration
-		let index = self.junctions.len().checked_sub(next_back)?;
-		if self.next > index {
-			return None
-		}
-		self.back = next_back;
-
-		self.junctions.at(index)
+		self.junctions.at(self.range.next_back()?)
 	}
 }
 
 impl<'a> IntoIterator for &'a Junctions {
 	type Item = &'a Junction;
 	type IntoIter = JunctionsRefIterator<'a>;
 	fn into_iter(self) -> Self::IntoIter {
-		JunctionsRefIterator { junctions: self, next: 0, back: 0 }
+		JunctionsRefIterator { junctions: self, range: 0..self.len() }
 	}
 }
 
 impl IntoIterator for Junctions {
 	type Item = Junction;
 	type IntoIter = JunctionsIterator;
 	fn into_iter(self) -> Self::IntoIter {
-		JunctionsIterator(self)
+		JunctionsIterator { range: 0..self.len(), junctions: self }
```

Comment:
would be nice if arg order was switched: junctions, range so it's the same as `Junctions RefIterator`

Reply:
This is deliberate. The arg order must be like this because `self` is moved into `JunctionsIterator`, and once you move it you can't call `self.len()`.